src/vos/vos_gc.c

```diff
@@ -0,0 +1,838 @@
+/**
+ * (C) Copyright 2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * This file is part of daos
+ *
+ * vos/vos_gc.c
+ */
+#define D_LOGFAC	DD_FAC(vos)
+
+#include <daos/btree.h>
+#include <daos/mem.h>
+#include <daos_srv/vos.h>
+#include "vos_internal.h"
+
+enum {
+	GC_CREDS_MIN	= 1,	/**< minimum credits for vos_gc_run */
+	GC_CREDS_PRIV	= 128,	/**< credits for internal usage */
+	GC_CREDS_MAX	= 4096,	/**< maximum credits for vos_gc_run */
+};
+
+/**
+ * Default garbage bag size
+ * - header of vos_gc_bag_df is 64 bytes
+ * - PMDK allocation overhead is 16 bytes,
```

Comment:
This comment is bit confusing, PMDK allocation overhead shouldn't affect bag size, if you allocate a 4k bag, you'll get full 4k space.

Reply:
I just don't want each bag consumes more than 4K physical space :-)

Reply:
Ah, I see. You actually didn't allocate 4k (but slightly less than 4k)

I don't think we should presume a PMDK allocation overhead (which really depends on PMDK internal implementation) and try to round-up allocation based on that in DAOS. Will this trick bring any benefit? (like reduce internal fragmentation?)

I think as long as the allocatioin size is smaller than 256k (huge chunk allocation in PMDK), there isn't much difference on different size allocations. Actually I'm afraid that such non 2^n size allocation could result in even more internal fragmentation.

In the future, I think we can use slab allocator for the bags (and the interior tree nodes), then there will be more freedom on choosing bag size.

Reply:
Yes, the actual bag size is slightly smaller than 4K, so even with the internal overhead of PMDK, it should fit into 4K, and we probably want to bump it a little bit, e.g. 8K or 16K.